sdk/core/core-http/lib/nodeFetchHttpClient.ts

```diff
@@ -22,8 +23,57 @@ if (typeof globalWithFetch.fetch !== "function") {
 }
 
 export class NodeFetchHttpClient extends FetchHttpClient {
+  private httpAgent: http.Agent | undefined;
+  private httpsAgent: https.Agent | undefined;
+  private httpClientOptions: HttpClientOptions;
+
   private readonly cookieJar = new tough.CookieJar(undefined, { looseMode: true });
 
+  constructor(httpClientOptions?: HttpClientOptions) {
+    super();
+
+    this.httpClientOptions = {
+      keepAlive: true,
+      proxySettings: undefined,
+      ...httpClientOptions
+    };
+  }
+
+  private getOrCreateAgent(httpRequest: WebResource): http.Agent | https.Agent {
+    let agent: http.Agent | https.Agent | undefined = undefined;
+    const isHttps = isUrlHttps(httpRequest.url);
+
+    if (isHttps) {
+      agent = this.httpsAgent;
+    } else {
+      agent = this.httpAgent;
+    }
+
+    if (!agent) {
+      if (this.httpClientOptions.proxySettings) {
+        const tunnel: ProxyAgent = createProxyAgent(
+          httpRequest.url,
+          this.httpClientOptions.proxySettings,
+          httpRequest.headers
+        );
+
+        agent = tunnel.agent;
+        if (tunnel.isHttps) {
+          this.httpsAgent = tunnel.agent as https.Agent;
+        } else {
+          this.httpAgent = tunnel.agent;
+        }
+      } else {
+        const agentOptions: http.AgentOptions | https.AgentOptions = {
+          keepAlive: this.httpClientOptions.keepAlive
```

Comment:
What do you think about having a default value for `maxSockets` if the user hasn't specified one to help make sure connections get re-used?

We can check the global agent 1st in case the user has changed the value there, and eventually I'd love if we exposed `maxSockets` as an option as well.

```js
const globalAgent = isHttps ? https.globalAgent : http.globalAgent;
// default maxSockets is Infinity
let maxSockets = 50; // our default?
if (globalAgent && globalAgent.maxSockets !== Infinity) {
  maxSockets = globalAgent.maxSockets;
}
```

Reply:
Yep, I plan to add a setting for maxSockets in another PR where I unify the `PipelineOptions` pattern we have floating around in a few libraries.

Regarding having a default for `maxSockets`, sounds good to me!  Maybe I'll also add that in the next PR since this one is mainly focused on agent reuse?

Reply:
Cool! I keep hounding on that point because my testing (admittedly over a year ago so things may have changed) showed that you don't really get the benefits of keepAlive with the default value maxSockets has today, but my info could be out of date :)  As long as it makes it in eventually I'll be happy!